exercises/binary-search-tree/binary_search_tree.exs

```diff
@@ -0,0 +1,27 @@
+defmodule BinarySearchTree do
+  @type bst_node :: %{data: any, left: bst_node | nil, right: bst_node | nil}
+
+  @doc """
+  Create a new Binary Search Tree with root's value as the given 'data'
+  """
+  @spec new(any) :: %{data: any, left: bst_node, right: bst_node}
```

Comment:
I think we need to update this typespec. In your implementation it returns a `bst_node` (since it's valid for the left and right nodes to be `nil` in that type definition) but this typespec implies that we have at least a balanced tree with a depth of 1.

Reply:
Good catch, I think I fixed this. Looks like I don't take advantage of my custom type like I did for insert. It can be simply defined as `@spec new(any):: bst_node` which for left and right is a union of another bst_node or nil.

In the case of a newly created node left and right are both nil.

In the case of an insert under an existing node, one of left or right will be non-nil(another bst_node)